### ISSUE TITLE
Migration to Symfony 4

### ISSUE THREAD

**Issue by chartjes** [opencfp/opencfp#618]
Silex is hitting EOL in June 2018, so I think it's time for us to figure out how to move OpenCFP more towards Symfony, perhaps targeting Symfony 4.

**Comment by mdwheele** [comment]
Laravel! :grinning:

**Comment by dbu** [comment]
i'd like to look into porting it to symfony 4 flex. no promises though, i'll try a bit and then report back if i think that i can seriously propose to do it, to check if that is whats wanted.

**Comment by chartjes** [comment]
@dbu I definitely think Symfony 4 is the target to move to given how much Symfony stuff we are already relying on. Would greatly appreciate you taking a look

**Comment by mdwheele** [comment]
Just as a heads up: I've started working on this. I've got the framework + some basic components (webpack-encore, twig, security) started. I'm working on pulling over our CI process and helping some things settle (the divergence between `.travis.yml`, `Makefile`, and `script/*`). I feel like TravisCI should be running (more or less) the same commands we are. If that happens to be `script/test`, cool. Currently, I've got to remember `make cs`, `script/setup` and `.travis.yml` is doing other things in `scripts`. Now's probably a good time to clean that up. 

Also pulling over all the `php-cs-fixer` / `.codeclimate.yml` stuff piece by piece.

I'm also including a simple LEMP stack on Docker that runs OpenCFP. You basically run `cfp shell` and it builds everything and drops you in a workspace to finish setting up the application.

I think that at *some point* it is going to be helpful to pump the breaks on PRs, but no need for that right now. I feel like the best option right now is to create a baseline, move over a few things, then we all agree to send PRs to a `symfony4` branch to move the rest of everything over. I think once I get the frontend working, complete with signup / login process handled, I'll put the branch up and we can all start migrating things in. 

This is also a GREAT time to create a decent service layer and inject dependencies into controllers rather than reaching out to the container constantly. Make it part of the "definition of done" for moving code into the new framework.

**Comment by dbu** [comment]
sounds like a great plan! and glad you are on it. i don't think i would really find time to work on it anytime soon, after all... too much interesting stuff to do, too little time.

**Comment by mdwheele** [comment]
That is exactly the problem I seem to find myself in pretty much all the time. Cheers!

**Comment by chartjes** [comment]
@mdwheele The Symfony project has a page where they trying to put together a small guide to help people migrate. Anything you run into that is difficult should be reported here.

https://github.com/symfony/symfony-docs/issues/8678

**Comment by mdwheele** [comment]
Hehe, I saw that! The migration is going well, so far. Our approach is very similar to others' in that issue. I'll make sure to report any problems I run into. I think we're actually going to be able to maintain backwards-compatibility, for what it's worth.

I'm scoping this migration to include account management, authentication and talk submissions. I am purposefully leaving out much of the administrative use-cases. This leaves us open to include an open-source bundle for that process or not. I think there are some biz discussions to be had.

It would be convenient to base this branch on a tagged release so I'm not having to manually patch in diffs from PRs to `upstream`. That said, it's going to be a bit longer before I'd say we need to do that so PRs away!

**Comment by derrabus** [comment]
I'm currently doing some research on how a migration path from Silex to Symfony could look like. Since this project is based on Silex and has a well maintained codebase, I gave it a shot. I even made it to a working prototype without the `silex/silex` package.

If you're interested, I can push it, but most tests are broken and the diff is too large to keep it in sync with changes that have already happened on master. Also, I didn't want to take away the fun of performing that migration. ;-)

If that is of any help and you're interested, I could pretty easily provide you with the following PRs:
* Controllers refactored to services. This makes then more or less autowirable when switching to Symfony. Also, the routing configuration remains compatible.
* Middlewares refactored to event subscribers. When migrating to Symfony, those can can be reused without modification.

Both refactorings are imho good preparations for a Symfony migration.

**Comment by mdwheele** [comment]
@derrabus if you want to push those changes to a branch on your fork, I'd be happy to take a look. I'm already pretty far into the move, myself, but would be interested in comparing. 

I wanted also to use this as an opportunity to clean up and not necessarily move everything over 1:1.

**Comment by derrabus** [comment]
@mdwheele Here you go: https://github.com/derrabus/opencfp/tree/remove-silex

**Comment by localheinz** [comment]
@derrabus 

Hehe, I remember your talk at SymfonyCon Berlin 2016: https://berlincon2016.symfony.com/speakers#session-1851. 

👍

**Comment by mdwheele** [comment]
Thanks @derrabus! With all that's going on in #850 (or in split PRs; whatever is decided), I think I'm going to hold off moving forward with my branch. If you want to send what you have as a PR, we could work through that. I get about 4-5 hours between 10pm - 3am to work on projects like this and have felt like the last few weeks have been 2 hours of rebasing to make 30 minutes of progress, haha. 

If others have time to carry this forward, I'd be happy to help review. If things settle a bit and the migration isn't resolved, I can pick it up then.

**Comment by derrabus** [comment]
@mdwheele I think this migration should not be done in a big-bang-all-or-nothing-PR. We can do this in small steps that all result in a runnable application. That's why I've submitted that PR: it reduces the coupling with Silex, so the diff for a Symfony 4 migration becomes smaller.

**Comment by mdwheele** [comment]
> I think this migration should not be done in a big-bang-all-or-nothing-PR. We can do this in small steps that all result in a runnable application. That's why I've submitted that PR: it reduces the coupling with Silex, so the diff for a Symfony 4 migration becomes smaller.

Agreed! Just to clarify; don't take my previous comment as "frustrated". It is what it is. I'll never ever say "too many contributors is a bad thing". I just don't want to get in the way :grinning:. We appreciate the help.

**Comment by derrabus** [comment]
A few things that need to be discussed for the migration:

* Symfony 4 requires php 7.1, OpenCFP currently requires 7.0. So your options are:
  * Go php 7.1 and Symfony 4.
  * Stay compatible to php 7.0 and live with Symfony 3.4 for now.
* Symfony compiles its DI container and router and it needs a place to dump them to. OpenCFP does have a cache directory, but it currently allows the user to disable cache usage. I would advise to remove that option.
* Symfony's framework configuration is usually written in YAML, so we need some place to store those configuration files. The default for a Symfony 4 application would be `/config`, but this directory is already taken for the application configuration. Since the user should not tinker with the framework configuration, I would not mix the files. My suggestiion would be `/resources/config`.

**Comment by dbu** [comment]
> we need some place to store those configuration files. 

one option might be to provide opencfp as a bundle that can be installed with flex. then you would have all system setup in the bundle and the installation would be to do start with symfony/skeleton and then `composer require opencfp/opencfp` and you can configure opencfp in the config file that is copied by flex to /config. makes the initial setup a little bit harder for non-symfony users, but maybe would be the easiest to maintain. also very easy to extend with own features for people who know symfony.

otherwise, i guess everybody would have to create some sort of fork of the opencfp repo to be able to use it.

**Comment by mdwheele** [comment]
> Symfony 4 requires php 7.1, OpenCFP currently requires 7.0. 

Think we were headed towards bumping to 7.1.12, specifically because of the changes in the internal server that happened in 7.1.12 and 7.2.0. It made it easier than having to support both sides of the fence. I think I echo @chartjes on "not *if* we go to Symfony 4; we *are* going to Symfony 4" so it looks like we're dropping support for PHP 7.0, which is good.

> Symfony compiles its DI container and router and it needs a place to dump them to. OpenCFP does have a cache directory, but it currently allows the user to disable cache usage. I would advise to remove that option.

I would personally prefer to let Symfony (and things like Twig) do what they were designed to do. I'd be :+1: 

> Symfony's framework configuration is usually written in YAML, so we need some place to store those configuration files. The default for a Symfony 4 application would be /config, but this directory is already taken for the application configuration. Since the user should not tinker with the framework configuration, I would not mix the files. My suggestiion would be /resources/config.

Is there an argument against providing all application config via `.env` in development (and actually environment variables in production)? There wouldn't be any problem moving `config` to achieve an incremental migration to Symfony. Or is that "just not how it's done"? It's confusing for me, as a newcomer, because the docs seem to suggest "PUT IT ALL IN .ENV. ITS THE SYMFONY WAY" and then certain values can't be read in config because of compile-time vs run-time.

**Comment by derrabus** [comment]
@dbu 

> one option might be to provide opencfp as a bundle that can be installed with flex.

Yes, that would be the way to go, if we don't see OpenCFP as a standalone application but as an extension to an existing Symfony application. Right now it's a standalone application. Changing this would be possible, but it's not a requirement for a Symfony migration.

> otherwise, i guess everybody would have to create some sort of fork of the opencfp repo to be able to use it.

No, you don't need to fork the project to use it now, so why would you need to fork it once Silex is gone? You can perfectly use the current configuration files to compile the container, like you would use a parameters.yml in Symfony 3.

@mdwheele

> Is there an argument against providing all application config via `.env` in development (and actually environment variables in production)?

You can do that, but again, you don't need to. If the current configuration file works for your users and the dev team, you can keep it that way.

> It's confusing for me, as a newcomer, because the docs seem to suggest "PUT IT ALL IN .ENV. ITS THE SYMFONY WAY" and then certain values can't be read in config because of compile-time vs run-time.

You're not alone in this. 😉  The Symfony 4 applications I've seen so far that have been migrated from Symfony 3 still use the old parameters.yml mechanism. If you're not in an all-Dockerized environment, you don't gain much from using environment variables, I guess.

**Comment by mdwheele** [comment]
> You can do that, but again, you don't need to. If the current configuration file works for your users and the dev team, you can keep it that way.

This is a good time for us to be reconsidering some of that. I think collapsing down to `.env` would be good for the project. We could wait on that though, like you say and do it later for 2.0.

> If you're not in an all-Dockerized environment, you don't gain much from using environment variables, I guess.

Oh but I want to be! Haha. I'm fleshing [this](https://github.com/mdwheele/opencfp-docker) out to the side for local development. For production installs not using Docker, vars can still be configured at the host level and a lot of PaaS will offer *something* to that regard to make it easier.

Somewhat unrelated... What actually *stops* you from using `.env` in production? With a correct server config, it's not servable (and in fact, I've seen a lot of shops specifically deny `.env` requests from untrusted zones).

**Comment by chartjes** [comment]
@mdwheele I am concerned that less-technically-inclined folks might mess things up at not configure things to deny .env requests. Again, we're trying to make sure people can checkout a release and install it with the fewest number of commands possible

**Comment by mdwheele** [comment]
The shops that have to put those rules in are already doing "weird things" by moving front controllers to the root of the project. Our own documentation about how to configure Apache is the "right way" to do it. 

```
<VirtualHost *:80>
    # Note document root is not where `.env` would be...
    DocumentRoot /path/to/web
    ServerName cfp.conference.com

    # Other Directives Here
</VirtualHost>
```

Ultimately, a `docker-compose.yml` that allows "pull this down and run `docker-compose up -d`" or Heroku is a good direction. At the end of the day though, someone is going to have to create an Apache / Nginx config if they want to host on their own boxes in production. That's what our users are doing today.

**Comment by dbu** [comment]
if everything that needs to be configured can be configured in parameters.yml / .env then indeed people don't need to fork.

as for exposing .env: normal webservers afaik never deliver hidden files (that start with `.`). plus, the .env file belongs into the root, and the web directory is the entry point for web requests. i think if you are afraid that people misconfigure their servers, config/parameters.yml is actually more dangerous than .env (because of at least having a unix-hidden file that some webservers won't deliver)

**Comment by chartjes** [comment]
@dbu Thanks for clarifying some things I was wondering about.

Again, as the least technical of the folks flooding this project with PR's I aim to keep getting OpenCFP installed and running as easy as possible. :)

**Comment by derrabus** [comment]
I think, we can keep the current configuration file as it is. That should keep the migration path easy. For the user, the only difference would be that the cache cannot be turned off anymore.

**Comment by derrabus** [comment]
I've made a new prototype. Have a look at #895, if you like.

**Comment by derrabus** [comment]
Seems like briannesbitt/Carbon#1028 is blocking Symfony 4 for us at the moment.

**Comment by BackEndTea** [comment]
On the topic of upgrading to php 7.1, It might be time to give all our dependencies a good look-over, and see if we want to swap any out, not all of them are being tested against php 7.1, some are not even being tested against php 7.0.

**Comment by derrabus** [comment]
#895 is ready for review.

**Comment by derrabus** [comment]
So, which libraries do currently block us from upgrading to Symfony 4?

```
$ composer why-not symfony/config:4
robmorgan/phinx  v0.9.1  requires  symfony/config (~2.8|~3.0)

$ composer why-not symfony/console:4
infection/infection  0.6.2   requires  symfony/console (^3.2)
robmorgan/phinx      v0.9.1  requires  symfony/console (~2.8|~3.0)

$ composer why-not symfony/finder:4
illuminate/filesystem  v5.5.17  requires  symfony/finder (~3.3)
infection/infection    0.6.2    requires  symfony/finder (^3.2)

$ composer why-not symfony/http-foundation:4
illuminate/validation  v5.5.17  requires  symfony/http-foundation (~3.3)

$ composer why-not symfony/process:4
infection/infection  0.6.2  requires  symfony/process (^3.2)

$ composer why-not symfony/translation:4
nesbot/carbon  1.22.1  requires  symfony/translation (~2.6 || ~3.0)

$ composer why-not symfony/yaml:4
infection/infection  0.6.2   requires  symfony/yaml (^3.3)
robmorgan/phinx      v0.9.1  requires  symfony/yaml (~2.8|~3.0)
```

**Comment by localheinz** [comment]
@derrabus 

What do you think about replacing `robmorgan/phinx` with `doctrine/migrations`?

This would also make sense in the light of switching to using `doctrine/orm`, if we wanted to.

**Comment by BackEndTea** [comment]
For infection we can move to the phar if needed. But i know they are working on being 4.0 compatible.

**Comment by localheinz** [comment]
How about removing all of the Laravel components?

**Comment by derrabus** [comment]
@localheinz Pick the library that works for you. Doctrine Migrations is powerful when used with the ORM. As a stand-alone tool, it's as good as any other migration tool, imho.

I wouldn't worry too much about blocking libraries just yet. Symfony 3.4 and 4.0 have the same feature set, so until May 2018 we would not gain much from upgrading to 4.0 anyway.

**Comment by derrabus** [comment]
Phinx will support Symfony 4 with its next release, see cakephp/phinx#1224.

**Comment by derrabus** [comment]
> How about removing all of the Laravel components?

I mean, you could replace Eloquent with Doctrine ORM, Sentinel with Symfony Security, Illuminate Validation with Symfony Validation etc. if you want to be more "standard Symfony". But as long as those libraries and are maintained and solve your problem, I don't see an urgent need to go that way.

**Comment by localheinz** [comment]
@derrabus 

For me it's not about being more *standard Symfony*. For me it's about being able to properly test components. If I see the effort required to test components using Eloquent models - or any other Active Record model, then I say "thanks, but no thanks".

**Comment by derrabus** [comment]
> If I see the effort required to test components using Eloquent models - or any other Active Record model, then I say "thanks, but no thanks".

That is a valid point indeed. Since this is imho not related to the Symfony migration, should we open another issue for this discussion?

**Comment by localheinz** [comment]
@derrabus

Sounds good! 👌

**Comment by derrabus** [comment]
There are now stable releases of Infection and Phinx that are compatible with Symfony 4. Now we just have to wait for the Illuminate components (or replace them).
